backends/cuda/ceed-cuda-qfunction.c

```diff
@@ -0,0 +1,179 @@
+  Ceed_Cuda *ceed_Cuda;
+  ierr = CeedGetData(ceed, (void*)&ceed_Cuda);
+  CeedInt numinputfields, numoutputfields;
+  ierr = CeedQFunctionGetNumArgs(qf, &numinputfields, &numoutputfields);
+  CeedChk(ierr);
+  const int blocksize = ceed_Cuda->optblocksize;
+
+  for (CeedInt i = 0; i < numinputfields; i++) {
+    ierr = CeedVectorGetArrayRead(U[i], CEED_MEM_DEVICE, &data->fields.inputs[i]);
+    CeedChk(ierr);
+  }
+
+  for (CeedInt i = 0; i < numoutputfields; i++) {
+    ierr = CeedVectorGetArray(V[i], CEED_MEM_DEVICE, &data->fields.outputs[i]);
+    CeedChk(ierr);
+  }
+
+  // TODO find a way to avoid this systematic memCpy
+
+  size_t ctxsize;
+  ierr = CeedQFunctionGetContextSize(qf, &ctxsize); CeedChk(ierr);
+  if (ctxsize > 0) {
+    if(!data->d_c) {
+      ierr = cudaMalloc(&data->d_c, ctxsize); CeedChk_Cu(ceed, ierr);
+    }
+    void *ctx;
+    bool fortranstatus;
+    ierr = CeedQFunctionGetFortranStatus(qf, &fortranstatus); CeedChk(ierr);
+    if (fortranstatus) {
+      ierr = CeedQFunctionGetFortranContext(qf, &ctx); CeedChk(ierr);
+    } else {
+      ierr = CeedQFunctionGetContext(qf, &ctx); CeedChk(ierr);
+    }
+    ierr = cudaMemcpy(data->d_c, ctx, ctxsize, cudaMemcpyHostToDevice); CeedChk_Cu(ceed, ierr);
+  }
+
+  void* ctx;
+  ierr = CeedQFunctionGetContext(qf, &ctx); CeedChk(ierr);
+  // void *args[] = {&ctx, (void*)&Q, &data->d_u, &data->d_v};
+  void *args[] = {&data->d_c, (void*)&Q, &data->fields};
+  ierr = run_kernel(ceed, data->qFunction, CeedDivUpInt(Q, blocksize), blocksize,
+                    args);
+  CeedChk(ierr);
+
+  for (CeedInt i = 0; i < numinputfields; i++) {
+    ierr = CeedVectorRestoreArrayRead(U[i], &data->fields.inputs[i]);
+    CeedChk(ierr);
+  }
+
+  for (CeedInt i = 0; i < numoutputfields; i++) {
+    ierr = CeedVectorRestoreArray(V[i], &data->fields.outputs[i]);
+    CeedChk(ierr);
+  }
+
+  return 0;
+}
+
+static int CeedQFunctionDestroy_Cuda(CeedQFunction qf) {
+  int ierr;
+  CeedQFunction_Cuda *data;
+  ierr = CeedQFunctionGetData(qf, (void*)&data); CeedChk(ierr);
+  Ceed ceed;
+  ierr = CeedQFunctionGetCeed(qf, &ceed); CeedChk(ierr);
+
+  CeedChk_Cu(ceed, cuModuleUnload(data->module));
+  ierr = cudaFree(data->d_c); CeedChk_Cu(ceed, ierr);
+
+  ierr = CeedFree(&data); CeedChk(ierr);
+
+  return 0;
+}
+
+static int loadCudaFunction(CeedQFunction qf, char* c_src_file) {
+  int ierr;
+  Ceed ceed;
+  CeedQFunctionGetCeed(qf, &ceed);
+  char* cuda_file;
+  ierr = CeedCalloc(CUDA_MAX_PATH, &cuda_file); CeedChk(ierr);
+  memcpy(cuda_file, c_src_file, strlen(c_src_file));
+  const char *last_dot = strrchr(cuda_file, '.');
+  if (!last_dot)
+    return CeedError(ceed, 1, "Cannot find file's extension!");
+  const size_t cuda_path_len = last_dot - cuda_file;
+  strcpy(&cuda_file[cuda_path_len], ".cu");
+  //*******************
+  FILE *fp;
+  long lSize;
+  char *buffer;
+
+  fp = fopen ( cuda_file, "rb" );
+  if( !fp ) perror(cuda_file),exit(1);
```

Comment:
Library code should never call `exit()`. Use `CeedError()`.

Reply:
Thanks.